proposals/lip-0049.md

```diff
@@ -223,135 +223,171 @@ When processing a cross-chain message `ccm`, the functions below may be called b
+            for each module mdl for which beforeCrossChainCommandExecution exists:
+                mdl.beforeCrossChainCommandExecution(ccu, ccm)
+            
+            # Execute the cross-chain command
+            crossChainCommand = cross-chain command associated with (ccm.moduleID, ccm.crossChainCommandID)
+            crossChainCommand.verify(ccm)
+            crossChainCommand.execute(ccm)
+
+            # Call the afterCrossChainCommandExecution functions from other modules
+            for each module mdl for which beforeCrossChainCommandExecution exists:
+                mdl.afterCrossChainCommandExecution(ccu, ccm)     
+
+            # Emit processing event. If an error occurs, this event emission is also undone.
+            emitEvent(
+                moduleID = MODULE_ID_INTEROPERABILITY,
+                typeID = TYPE_ID_CCM_PROCESSED,
+                data = { "ccmID": ccmID, "status": CCM_PROCESSED_STATUS_COMMAND_APPLIED},
+                topics = [ccm.sendingChainID, ccm.receivingChainID]
+            )
+        except:
+            revert state to snapshot 
+            emitEvent(
+                moduleID = MODULE_ID_INTEROPERABILITY,
+                typeID = TYPE_ID_CCM_PROCESSED,
+                data = { "ccmID": ccmID, "status": CCM_PROCESSED_STATUS_COMMAND_FAILED},
+                topics = [ccm.sendingChainID, ccm.receivingChainID]
+            )  
             return
 
-    # Apply the message if the logic is supported and if the sending chain is not terminated
-    if (ccm.moduleID, ccm.crossChainCommandID) is supported:
-        call the logic associated with (ccm.moduleID, ccm.crossChainCommandID) on ccm
-
-    elif ccm.status != CCM_STATUS_OK or ccm.fee < MIN_RETURN_FEE*size(ccm):
-        ccm is discarded and has no further effect
-
-    elif moduleID is not supported:
-        sendInternal(
-            EMPTY_FEE_ADDRESS,
-            ccm.moduleID,
-            ccm.crossChainCommandID,
-            ccm.sendingChainID,
-            0,
-            CCM_STATUS_MODULE_NOT_SUPPORTED,
-            ccm.params
+    # If the status is not ok and the fee is sufficient, we send back an error message.
+    elif ccm.status == CCM_STATUS_OK and ccm.fee >= MIN_RETURN_FEE*size(ccm):
+        if moduleID is not supported:
+            errorStatus = CCM_STATUS_MODULE_NOT_SUPPORTED
+            eventStatus = CCM_PROCESSED_STATUS_MODULE_NOT_SUPPORTED
+        elif crossChainCommandID is not supported:
+            errorStatus = CCM_STATUS_CROSS_CHAIN_COMMAND_NOT_SUPPORTED
+            eventStatus = CCM_PROCESSED_STATUS_COMMAND_NOT_SUPPORTED
+
+        emitEvent(
+            moduleID = MODULE_ID_INTEROPERABILITY,
+            typeID = TYPE_ID_CCM_PROCESSED,
+            data = { "ccmID": ccmID, "status": eventStatus},
+            topics = [ccm.sendingChainID, ccm.receivingChainID]
         )
 
-    elif crossChainCommandID is not supported:
-        sendInternal(
+        send(
```

Comment:
Why not use the `error` method here?

Reply:
before error was only a method (not an internal function). I think it wouldn't change anything, right?

Reply:
Now we uniformly use the `bounce` function which is also nice. Can be resolved then.